scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -1732,41 +1700,66 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
       }
     }
 
-  private def convertToParam(tree: Term): Option[Term.Param] = {
-    def getType: Option[Type] = tree match {
-      case t: Term.Ascribe => Some(t.tpe)
+  private def convertToParam(tree: Tree): Option[Term.Param] = {
+    def getModFromName(name: Name): Option[Mod] = name.value match {
+      case soft.KwUsing() => Some(copyPos(name)(Mod.Using()))
       case _ => None
     }
-    @tailrec def getName(t: Tree, nest: Int = 0): Option[Name] = t match {
-      case t: Name => Some(t)
-      case t: Quasi => Some(t.become[Term.Name])
-      case t: Term.Select => Some(t.name)
-      case t: Term.Placeholder => Some(copyPos(t)(Name.Placeholder()))
-      case t: Term.Eta => Some(atPos(t.endTokenPos)(Name.Placeholder()))
-      case t: Term.Ascribe if nest == 0 => getName(t.expr, 1)
+    def getMod(t: Tree, mods: List[Mod] = Nil): Option[List[Mod]] = t match {
+      case t: Term.Name => getModFromName(t).map(_ :: mods)
       case _ => None
     }
-    @tailrec def getMod(t: Tree, nest: Int = 0): List[Mod.ParamsType] = t match {
-      case _: Quasi => Nil
-      case t: Term.Ascribe if nest == 0 => getMod(t.expr, 1)
-      case t: Term.Select if nest <= 1 => getMod(t.qual, 2)
-      case t: Term.Eta => getMod(t.expr, 2)
-      case t @ Term.Name(soft.KwUsing.name) => copyPos(t)(Mod.Using()) :: Nil
-      case _ => Nil
+    def getNameAndMod(t: Tree): Option[(Name, List[Mod])] = t match {
+      case t: Name => Some((t, Nil))
+      case t: Quasi => Some((t.become[Term.Name], Nil))
+      case t: Term.Select => getMod(t.qual).map((t.name, _))
+      case t: Term.Placeholder => Some((copyPos(t)(Name.Placeholder()), Nil))
+      case t: Term.Eta => getMod(t.expr).map((atPos(t.endTokenPos)(Name.Placeholder()), _))
+      case _ => None
     }
+
     tree match {
       case q: Quasi => Some(q.become[Term.Param])
       case _: Lit.Unit => None
+      case t: Term.Ascribe =>
+        getNameAndMod(t.expr).map { case (name, mod) =>
+          copyPos(t)(Term.Param(mod, name, Some(t.tpe), None))
+        }
       case t =>
-        val name = getName(t).getOrElse(syntaxError("not a legal formal parameter", at = t))
-        Some(copyPos(t)(Term.Param(getMod(t), name, getType, None)))
+        getNameAndMod(t).map { case (name, mod) =>
+          copyPos(t)(Term.Param(mod, name, None, None))
+        }
     }
   }
 
-  private def convertToParamClause(tree: Term): Term.ParamClause = (tree match {
-    case Term.Tuple(ts) => ts.flatMap(convertToParam)
-    case _ => convertToParam(tree).toList
-  }).reduceWith(toParamClause(None))
+  private def convertToParamClause(tree: Term)(
+      okNameLike: => Boolean,
+      okParamLike: => Boolean
```

Comment:
```suggestion
      isNameLike: => Boolean,
      isParamLike: => Boolean
```
?

Reply:
i think the meaning here is not "is this a name" but "is a name allowed here". should i rename to `isNameAllowed"?

Reply:
I think that would be clearer, thanks!

Reply:
done

Reply:
Thanks!